File: android/src/main/java/expo/modules/arcgis/GeoprocessingFunctions.kt
package expo.modules.arcgis

import com.arcgismaps.data.FeatureCollectionTable
import com.arcgismaps.mapping.view.Graphic
import com.arcgismaps.tasks.geoprocessing.GeoprocessingTask
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingBoolean
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingDouble
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingFeatures
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingLinearUnit
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingLong
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingParameter
import com.arcgismaps.tasks.geoprocessing.geoprocessingparameters.GeoprocessingString
import java.util.concurrent.ConcurrentHashMap

/**
 * Free functions backing the JS `geoprocessor` namespace — runs a geoprocessing service via a
 * [GeoprocessingTask]. Registered as an `AsyncFunction` in `ExpoArcgisGeometryModule`.
 */

/** Caches one [GeoprocessingTask] per service URL so repeated runs don't reload service metadata. */
private val geoprocessingTasks = ConcurrentHashMap<String, GeoprocessingTask>()

private fun geoprocessingTask(url: String): GeoprocessingTask =
  geoprocessingTasks.getOrPut(url) { GeoprocessingTask(url) }

internal suspend fun executeGeoprocessing(
  serviceUrl: String,
  inputs: Map<String, Any?>,
): Map<String, Any?> {
  val task = geoprocessingTask(serviceUrl)
  // createDefaultParameters() loads the task and sets the service's execution type (sync / async).
  val parameters = task.createDefaultParameters().getOrThrow()
  inputs.forEach { (name, raw) ->
    (raw as? Map<*, *>)?.let { buildGeoprocessingParameter(it) }?.let { parameters.inputs[name] = it }
  }
  val job = task.createJob(parameters)
  job.start()
  val result = job.result().getOrThrow()
  return mapOf("outputs" to serializeOutputs(result.outputs))
}

private fun buildGeoprocessingParameter(d: Map<*, *>): GeoprocessingParameter? = when (d["type"]) {
  "string" -> GeoprocessingString(d["value"] as? String ?: "")
  "double" -> GeoprocessingDouble((d["value"] as? Number)?.toDouble() ?: 0.0)
  "long" -> GeoprocessingLong((d["value"] as? Number)?.toInt() ?: 0)
  "boolean" -> GeoprocessingBoolean(d["value"] as? Boolean ?: false)
  "linearUnit" -> GeoprocessingLinearUnit(
    (d["value"] as? Number)?.toDouble() ?: 0.0,
    linearUnit(d["unit"] as? String),
  )
  "features" -> {
    val graphics = (d["geometries"] as? List<*>)
      ?.mapNotNull { (it as? Map<*, *>)?.let { m -> geometryFromDict(m) } }
      ?.map { Graphic().apply { geometry = it } } ?: emptyList()
    GeoprocessingFeatures(FeatureCollectionTable(graphics, emptyList()))
  }
  else -> null
}

private suspend fun serializeOutputs(outputs: Map<String, GeoprocessingParameter>): Map<String, Any?> {
  val result = mutableMapOf<String, Any?>()
  for ((name, param) in outputs) result[name] = serializeGeoprocessingParameter(param)
  return result
}

private suspend fun serializeGeoprocessingParameter(param: GeoprocessingParameter): Any? = when (param) {
  is GeoprocessingString -> param.value
  is GeoprocessingDouble -> param.value
  is GeoprocessingLong -> param.value
  is GeoprocessingBoolean -> param.value
  is GeoprocessingLinearUnit -> param.distance
  is GeoprocessingFeatures -> {
    if (param.canFetchOutputFeatures) param.fetchOutputFeatures().getOrThrow()
    param.features?.map { serializeFeature(it) } ?: emptyList<Map<String, Any?>>()
  }
  else -> null
}

File: ios/GeoprocessingFunctions.swift
import ArcGIS
import Foundation

/// Free functions backing the JS `geoprocessor` namespace — runs a geoprocessing service via a
/// `GeoprocessingTask`. Registered as an `AsyncFunction` in `ExpoArcgisGeometryModule`.

/// Caches one `GeoprocessingTask` per service URL so repeated runs don't reload service metadata.
private final class GeoprocessingTaskCache: @unchecked Sendable {
  static let shared = GeoprocessingTaskCache()
  private let lock = NSLock()
  private var tasks: [String: GeoprocessingTask] = [:]

  func task(for url: String) -> GeoprocessingTask? {
    lock.lock()
    defer { lock.unlock() }
    if let task = tasks[url] { return task }
    guard let parsed = URL(string: url) else { return nil }
    let task = GeoprocessingTask(url: parsed)
    tasks[url] = task
    return task
  }
}

func executeGeoprocessing(_ serviceUrl: String, _ inputs: [String: Any]) async throws -> JobRef {
  guard let task = GeoprocessingTaskCache.shared.task(for: serviceUrl) else { throw OfflineError.invalidParameters }
  // makeDefaultParameters() loads the task and sets the service's execution type (sync / async).
  let parameters = try await task.makeDefaultParameters()
  for (name, raw) in inputs {
    if let dict = raw as? [String: Any], let param = buildGeoprocessingParameter(dict) {
      parameters.setInputValue(param, forKey: name)
    }
  }
  let job = task.makeJob(parameters: parameters)
  return JobRef(job: job) {
    let result = try await job.result.get()
    return ["outputs": try await serializeOutputs(result.outputs)]
  }
}

private func buildGeoprocessingParameter(_ d: [String: Any]) -> GeoprocessingParameter? {
  switch d["type"] as? String {
  case "string":
    return GeoprocessingString(value: d["value"] as? String ?? "")
  case "double":
    return GeoprocessingDouble(value: (d["value"] as? NSNumber)?.doubleValue ?? 0)
  case "long":
    return GeoprocessingLong(value: (d["value"] as? NSNumber)?.int32Value ?? 0)
  case "boolean":
    return GeoprocessingBoolean(value: d["value"] as? Bool ?? false)
  case "date":
    return GeoprocessingDate(value: Date(timeIntervalSince1970: ((d["value"] as? NSNumber)?.doubleValue ?? 0) / 1000))
  case "linearUnit":
    return GeoprocessingLinearUnit(
      distance: (d["value"] as? NSNumber)?.doubleValue ?? 0,
      linearUnit: linearUnit(d["unit"] as? String)
    )
  case "features":
    let geometries = (d["geometries"] as? [[String: Any]] ?? []).compactMap(geometryFromDict)
    let table = FeatureCollectionTable(geoElements: geometries.map { Graphic(geometry: $0) }, fields: [])
    return GeoprocessingFeatures(features: table)
  case "multiValue":
    let rawValues = d["values"] as? [Any] ?? []
    // Build elements: JS numbers → GeoprocessingDouble, JS strings → GeoprocessingString.
    // NSNumber coming from the bridge is used for both; Bool is excluded (its objCType is "c").
    let elements: [GeoprocessingParameter] = rawValues.compactMap { v in
      if let n = v as? NSNumber, !(v is Bool), String(cString: n.objCType) != "c" {
        return GeoprocessingDouble(value: n.doubleValue)
      } else if let s = v as? String {
        return GeoprocessingString(value: s)
      }
      return nil
    }
    // `parameterType` is the metatype of the element kind; fall back to GeoprocessingString.
    let paramType: GeoprocessingParameter.Type = elements.first is GeoprocessingDouble
      ? GeoprocessingDouble.self
      : GeoprocessingString.self
    return GeoprocessingMultiValue(parameterType: paramType, values: elements)
  case "dataFile":
    guard let urlStr = d["url"] as? String, let url = URL(string: urlStr) else { return nil }
    return GeoprocessingDataFile(url: url)
  default:
    return nil
  }
}

private func serializeOutputs(_ outputs: [String: GeoprocessingParameter]) async throws -> [String: Any] {
  var result: [String: Any] = [:]
  for (name, param) in outputs {
    result[name] = try await serializeGeoprocessingParameter(param)
  }
  return result
}

private func serializeGeoprocessingParameter(_ param: GeoprocessingParameter) async throws -> Any {
  switch param {
  case let p as GeoprocessingString: return p.value
  case let p as GeoprocessingDouble: return p.value
  case let p as GeoprocessingLong: return Int(p.value)
  case let p as GeoprocessingBoolean: return p.value
  case let p as GeoprocessingDate: return p.value.timeIntervalSince1970 * 1000
  case let p as GeoprocessingLinearUnit: return p.distance
  case let p as GeoprocessingFeatures:
    if p.canFetchOutputFeatures { try await p.fetchOutputFeatures() }
    return p.features?.features().map(serializeFeature) ?? []
  default:
    return NSNull()
  }
}
